subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java

```diff
@@ -164,6 +164,11 @@ public IncrementalDetectedExecutionStateChanges(
         public InputChangesInternal createInputChanges() {
             return new IncrementalInputChanges(inputFileChanges, incrementalInputProperties);
         }
+
+        @Override
+        public boolean isIncremental() {
```

Comment:
I think `incremental` is a sufficiently overloaded term that we should qualify what we mean by it here, so call this maybe `isIncrementalExecutionAllowed()` or something.

However, I think there's a code smell here: we have a type hierarchy where everyone returns `false` for `isX()`, except for a single implementation called `XImpl`. I think it might be better to move the functionality dependent on `isIncremental()` to the implementation here, and move `logExecutionReasons()` to `ExexcutionStateChanges`. WDYT?